File: backend/app/tables/knowledge.py
import json

def init_knowledge_docs(conn):
    """保留旧表兼容"""
    conn.execute("""
        CREATE TABLE IF NOT EXISTS knowledge_docs (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            title TEXT NOT NULL,
            content TEXT NOT NULL,
            category TEXT DEFAULT '未分类',
            tags TEXT DEFAULT '[]',
            source TEXT DEFAULT '',
            doc_type TEXT DEFAULT 'text',
            user_id INTEGER NOT NULL,
            base_id INTEGER DEFAULT 0,
            is_active INTEGER DEFAULT 1,
            created_at TEXT DEFAULT (datetime('now')),
            updated_at TEXT DEFAULT (datetime('now')),
            FOREIGN KEY (user_id) REFERENCES users(id)
        );
    """)

def init_knowledge_items(conn):
    """CherryStudio 对齐的 knowledge_items 表"""
    conn.execute("""
        CREATE TABLE IF NOT EXISTS knowledge_items (
            id TEXT PRIMARY KEY,
            base_id TEXT NOT NULL,
            group_id TEXT DEFAULT NULL,
            type TEXT NOT NULL CHECK(type IN ('file', 'url', 'note', 'directory')),
            data TEXT NOT NULL DEFAULT '{}',
            status TEXT NOT NULL DEFAULT 'idle'
                CHECK(status IN ('idle', 'preparing', 'processing', 'reading', 'embedding', 'completed', 'failed', 'deleting')),
            error TEXT DEFAULT NULL,
            created_at TEXT DEFAULT (datetime('now')),
            updated_at TEXT DEFAULT (datetime('now')),
            FOREIGN KEY (base_id) REFERENCES knowledge_bases(id) ON DELETE CASCADE
        );
    """)

    conn.execute("""
        CREATE INDEX IF NOT EXISTS idx_knowledge_items_base_type_created
        ON knowledge_items(base_id, type, created_at);
    """)

    conn.execute("""
        CREATE INDEX IF NOT EXISTS idx_knowledge_items_base_group_created
        ON knowledge_items(base_id, group_id, created_at);
    """)

    conn.execute("""
        CREATE TABLE IF NOT EXISTS knowledge_groups (
            id TEXT PRIMARY KEY,
            user_id INTEGER NOT NULL,
            name TEXT NOT NULL,
            created_at TEXT DEFAULT (datetime('now')),
            updated_at TEXT DEFAULT (datetime('now')),
            FOREIGN KEY (user_id) REFERENCES users(id) ON DELETE CASCADE
        );
    """)

File: frontend/src/types/knowledge.ts
export type KnowledgeItemType = 'file' | 'url' | 'note' | 'directory'
export type KnowledgeItemStatus = 'idle' | 'preparing' | 'processing' | 'reading' | 'embedding' | 'completed' | 'failed' | 'deleting'
export type KnowledgeSearchMode = 'default' | 'bm25' | 'hybrid'
export type KnowledgeBaseStatus = 'completed' | 'failed'

export interface KnowledgeBase {
  id: string
  name: string
  groupId: string | null
  dimensions: number | null
  embeddingModelId: string | null
  status: KnowledgeBaseStatus
  error: string | null
  rerankModelId: string | null
  fileProcessorId: string | null
  chunkSize: number
  chunkOverlap: number
  threshold: number | null
  documentCount: number | null
  searchMode: KnowledgeSearchMode
  hybridAlpha: number | null
  createdAt: string
  updatedAt: string
  itemCount?: number
}

export interface KnowledgeBaseListItem extends KnowledgeBase {
  itemCount: number
}

export interface FileItemData {
  source: string
  fileEntryId?: string
}

export interface UrlItemData {
  source: string
  url: string
}

export interface NoteItemData {
  source: string
  content: string
  sourceUrl?: string
}

export interface DirectoryItemData {
  source: string
  path: string
}

export type KnowledgeItemData = FileItemData | UrlItemData | NoteItemData | DirectoryItemData

export interface KnowledgeItem {
  id: string
  baseId: string
  groupId: string | null
  type: KnowledgeItemType
  data: KnowledgeItemData
  status: KnowledgeItemStatus
  error: string | null
  createdAt: string
  updatedAt: string
}

export interface KnowledgeChunkMetadata {
  itemId: string
  itemType: KnowledgeItemType
  source: string
  chunkIndex: number
  tokenCount: number
}

export interface KnowledgeSearchResult {
  pageContent: string
  score: number
  scoreKind: 'relevance' | 'ranking'
  rank: number
  metadata: KnowledgeChunkMetadata
  itemId?: string
  chunkId: string
}

export interface KnowledgeItemChunk {
  id: string
  itemId: string
  content: string
  metadata: KnowledgeChunkMetadata
}

export interface KnowledgeGroup {
  id: string
  name: string
  userId?: number
  createdAt?: string
  updatedAt?: string
}

export interface OffsetPaginationResponse<T> {
  items: T[]
  total: number
  page: number
}

export type KnowledgeTabKey = 'file' | 'note' | 'directory' | 'url'
